src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,416 @@
-        match self.mdaa.last_updated.cmp(&self.mdab.last_updated) {
-            Ordering::Less => &self.mdab,
-            Ordering::Greater => &self.mdaa,
-            _ => &self.mdab,
+        let crc = crc32::checksum_ieee(&buf[4..SECTOR_SIZE as usize]);
+        if crc != LittleEndian::read_u32(&buf[..4]) {
+            return Err(EngineError::Engine(ErrorEnum::Invalid, "header CRC invalid".into()));
         }
+
+        let blkdev_size = Sectors(LittleEndian::read_u64(&buf[20..28]) as u64);
+
+        let pool_uuid = try!(Uuid::parse_str(from_utf8(&buf[32..64]).unwrap())
+            .map_err(|_| EngineError::Engine(ErrorEnum::Invalid, "invalid pool uuid".into())));
+        let dev_uuid = try!(Uuid::parse_str(from_utf8(&buf[64..96]).unwrap())
+            .map_err(|_| EngineError::Engine(ErrorEnum::Invalid, "invalid dev uuid".into())));
+
+
+        let mda_size = Sectors(LittleEndian::read_u64(&buf[96..104]) as u64);
+
+        try!(validate_mda_size(mda_size));
+
+        Ok(StaticHeader {
+            pool_uuid: pool_uuid,
+            dev_uuid: dev_uuid,
+            blkdev_size: blkdev_size,
+            mda_size: mda_size,
+            reserved_size: Sectors(LittleEndian::read_u64(&buf[104..112]) as u64),
+            flags: 0,
+        })
+    }
+
+    /// Read pool uuid from buf at offset.
+    /// Return an error if the data can't be interpreted as a Uuid.
+    pub fn read_pool_uuid(buf: &[u8; BDA_STATIC_HDR_SIZE as usize]) -> EngineResult<Uuid> {
+        Uuid::parse_str(from_utf8(&buf[SECTOR_SIZE as usize + 32..SECTOR_SIZE as usize + 64])
+                .unwrap())
+            .map_err(|_| EngineError::Engine(ErrorEnum::Invalid, "invalid pool uuid".into()))
     }
 
-    /// Return the least recently timestamped MDA.
-    pub fn least_recent(&mut self) -> &mut MDA {
-        match self.mdaa.last_updated.cmp(&self.mdab.last_updated) {
-            Ordering::Less => &mut self.mdaa,
-            Ordering::Greater => &mut self.mdab,
-            _ => &mut self.mdaa,
+    /// Determine the ownership of a device based on data in buf.
+    pub fn determine_ownership(buf: &[u8; BDA_STATIC_HDR_SIZE as usize])
+                               -> EngineResult<DevOwnership> {
+        if &buf[SECTOR_SIZE as usize + 4..SECTOR_SIZE as usize + 20] == STRAT_MAGIC {
+            Ok(DevOwnership::Ours(try!(StaticHeader::read_pool_uuid(&buf))))
+        } else {
+            if buf.iter().any(|x| *x != 0) {
+                Ok(DevOwnership::Theirs)
+            } else {
+                Ok(DevOwnership::Unowned)
+            }
         }
     }
 }
 
 #[derive(Debug, Clone)]
-pub struct SigBlock {
-    pub total_size: Sectors,
-    pub pool_uuid: Uuid,
-    pub dev_uuid: Uuid,
-    pub mda: MDAGroup,
-    pub mda_sectors: Sectors,
-    pub reserved_sectors: Sectors,
+pub struct MDARegions {
+    // Spec defines 4 regions, but regions 2 & 3 are duplicates of 0 and 1 respectively
+    pub region_size: Sectors,
+    pub mdas: [MDAHeader; 2],
```

Comment:
This literal 2 should have a name and be derived from NUM_MDA_REGIONS or possibly NUM_MDA_REGIONS should be derived from the name for this 2.